### ISSUE TITLE
[question] Conflict between dependencies

### ISSUE THREAD

**Issue by Alberto-Izquierdo** [conan-io/conan-center-index#22203]
### What is your question?

Hi, I have a question regarding conflicts between dependencies.

My conanfile.py looks like this:

```
class MyProjectConan(ConanFile):
    requires = [
        "nlohmann_json/3.11.2",
        "jinja2cpp/1.2.1",
        "json-schema-validator/2.3.0",
        ...
    ]
    ...
    default_options = {
        "shared": False,
        ...
    }
```

Because [one of the changes](https://github.com/conan-io/conan-center-index/pull/21557/files) in the json-schema-validator, when I try to run conan install, I get this error:

```
ERROR: Version conflict: json-schema-validator/2.3.0->nlohmann_json/3.11.3, X/X->nlohmann_json/3.11.2.
```

I have tried using ranges to specify the nlohmann_json's version, using the default_options to force (`"nlohmann_json/*:force": True`) the version as specified in the [documentation](https://docs.conan.io/2.0/tutorial/versioning/conflicts.html), and using the shared option (`"nlohmann_json/*:shared": True`), but none of them works. Is there something I am missing?

Would changing the `json-schema-validator` recipe to use a range instead of a fixed version be an appropriate solution? Because I can create a PR with that change.

**Comment by perseoGI** [comment]
Greetings @Alberto-Izquierdo (btw we both have the same surname!)

Using the `force` trait is not an elegant solution.

We recommend using version ranges, but version ranges should be used carefully and **order matters**!

For example, in your recipe:

```
self.requires("nlohmann_json/3.11.2")
self.requires("json-schema-validator/2.3.0")
```
This will definitely generate an error: 
```
ERROR: Version conflict: Conflict between nlohmann_json/3.11.3 and nlohmann_json/3.11.2 in the graph.
Conflict originates from json-schema-validator/2.3.0
```

In this case, as `json-schema-validator/2.3.0` currently depends on a fixed `nlohmann_json` version (`3.11.2`), an easy approach could be to first depend on `json-schema-validator` and later, depend on `nlohmann_json` with a version range.

```
self.requires("json-schema-validator/2.3.0")
self.requires("nlohmann_json/[*]")
```
Conan analyzes and computes de dependency graph in order.
This means that Conan will first determine that your project depends on `json-schema-validator/2.3.0`, then, it will look up its dependencies, finding `nlohmann_json/3.11.2`.

This way, when Conan finds `nlohmann_json/[*]`, as this is the most "laxe" version range, it will resolve to the current project dependency (`nlohmann_json/3.11.2`).

---

Notice that if you consider putting it the other way, like:
```
self.requires("nlohmann_json/[*]")
self.requires("json-schema-validator/2.3.0")
```
Conan will first resolve to the latest `nlohmann_json` version (currently `3.12.0`) and you will get a conflict error.

---

I'm closing this issue, expecting to solve your doubts. Anyways, we are currently trying to use more version ranges across Conan Center Index to avoid this kind of conflict, but sometimes, they are inevitable...
Do not hesitate to ask any other doubts here or open a new ticket if unrelated. Thank you!

**TLDR: use version ranges and order matters when declaring project dependencies!**
